core/src/main/java/hudson/model/AsyncAperiodicWork.java

```diff
@@ -91,8 +138,39 @@ public void run() {
     }
 
     protected StreamTaskListener createListener() {
+        File f = getLogFile();
+        if (f.isFile()) {
+            if ((lastRotateMillis == null)
+                    || (lastRotateMillis + logRotateMillis < System.currentTimeMillis())
+                    || (logRotateSize > 0 && f.length() > logRotateSize)) {
+                lastRotateMillis = System.currentTimeMillis();
+                File p = null;
+                for (int i = 5; i >= 0; i--) {
+                    File o = i == 0 ? f : new File(f.getParentFile(), f.getName() + "." + i);
+                    if (o.isFile()) {
+                        if (p != null && !p.exists()) {
+                            if (!o.renameTo(p)) {
+                                logger.log(getErrorLoggingLevel(), "Could not rotate log files {0} to {1}",
+                                        new Object[]{o, p});
+                            }
+                        } else {
+                            if (!o.delete()) {
+                                logger.log(getErrorLoggingLevel(), "Could not delete log files {0} to enable rotation",
+                                        o);
+                            }
+                        }
+                        p = o;
+                    }
```

Comment:
so first iteration:
`i = 5` `o=fileName5` `p = null` --> `o` gets deleted, `p = fileName5` now and does not exist cause it was deleted

second iteration:
`i = 4` `o=fileName4` `p=fileName5` and `p` does not exist --> `o` gets renamed to `p` = `fileName5` then `p` is set to `o` which is `fileName5`

shouldn't be `p = fileName4` at this point?